cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+type RolloutStatus int
+
+const (
+	// resource desired status is not applied yet
+	ToApply RolloutStatus = iota
+	// resource desired status is applied and last applied status is not updated
+	Progressing
+	// resource desired status is applied and last applied status is succeed
+	Succeed
+	// resource desired status is applied and last applied status is failed
+	Failed
+	// when the rollout status is progressing or failed and the status remains for longer than the timeout
+	// then the status will be set to timeout
+	TimeOut
+	// skip rollout on this cluster
+	Skip
+)
+
+// Return the rollout status on each managed cluster
+type ClusterRolloutStatusFunc func(clusterName string) ClusterRolloutStatus
+
+type ClusterRolloutStatus struct {
+	// cluster name
+	// required field
+	Name string
+	// cluster group key
+	// optional field
+	GroupKey clusterv1beta1.GroupKey
+	// rollout status
+	// required field
+	Status RolloutStatus
+	// the last transition time of rollout status
+	// this is used to calculate timeout for progressing and failed status
+	// optional field
+	LastTransitionTime *time.Time
+}
+
+type RolloutResult struct {
+	// clusters to rollout
+	ClustersToRollout map[string]ClusterRolloutStatus
+	// clusters that are timeout
+	ClustersTimeOut map[string]ClusterRolloutStatus
+}
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// placement decision tracker
+	pdTracker *clusterv1beta1.PlacementDecisionClustersTracker
+}
+
+func NewRolloutHandler(pdTracker *clusterv1beta1.PlacementDecisionClustersTracker) *RolloutHandler {
+	if pdTracker == nil {
+		return nil
+	}
+
+	return &RolloutHandler{
+		pdTracker: pdTracker,
+	}
+}
+
+// The input is a duck type RolloutStrategy and a ClusterRolloutStatusFunc to return the rollout status on each managed cluster.
+// Return the strategy actual take effect and a list of cluster that needs to update with current state.
+func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	switch rolloutStrategy.Type {
```

Comment:
What's the behavior if the rollout strategy changes in the middle of a rollout?

Reply:
When the rollout strategy changes, it should trigger another reconcile to call GetRolloutCluster() I think. The returned clusters to rollout based on the current status snapshot.

Reply:
one thing may not clear here, if a cluster removed/added from the placement (even without changing the  rolloutStrategy) the consumer controller should track this behavior from placementDecisionTracker [here](https://github.com/open-cluster-management-io/api/pull/254/files#diff-a55f54ab06035f29ac70ef3b6c049983eef5c90f296007e3a6b6d6ca171f73b0R149)  
Is that correct ?

Reply:
Not sure where "here" points to (the URL doesn't work), the consumer controller needs to run pdTracker.Get() to refresh the placement decisions and will also get the added and deleted clusters.

Reply:
yap, the URL pointing to the pdTracker.Get() function.

Reply:
Say for example the mandatory decision group completes and the controller has moved on to the next group (we'll call it group2). Then a new cluster is added to the mandatory decision group, `GetRolloutCluster` would return just that single cluster that was added and the controller would ignore group2 until that new cluster finishes?

If that's the case, what happens to the partial rollout in `group2` if that new cluster in the mandatory decision group does not succeed?

Reply:
> GetRolloutCluster would return just that single cluster that was added and the controller would ignore group2 until that new cluster finishes?

Yes, I think so.

> If that's the case, what happens to the partial rollout in group2 if that new cluster in the mandatory decision group does not succeed?

If group2 is partially rollout and the new cluster fails, the group2 rollout progress won't continue (clusters already upgraded won't rollback) until the new cluster succeeds.